File: forage.py
# importing libraries

import sys # this lets us receive passed arguments in the command line
import getopt # this lets us receive arguments and options in the command line
import json # this lets us parse JSON files
import csv # this lets us parse CSV files
import dateutil.parser # this helps us parse dates
import calendar # this helps us understand dates
import requests # this helps us make good HTTP requests
from contextlib import closing # this helps us make good HTTP requests

# Setting a few variables

output = {}
data = {}
write = False
hydrometric = False

# Here, we get the arguments and parameters passed to our script, and do things accordingly!

try:
    opts, args = getopt.getopt(sys.argv[1:], 'w:h', ['write=', 'help'])
except getopt.GetoptError:
    print "Uh oh, something went wrong. Try again, and read the docs! Use forage.py -h if you need help."
    sys.exit(2)

for opt, arg in opts:
    if opt in ('-h', '--help'):
        print "Help docs coming soon!"
        sys.exit(2)
    elif opt in ('-w', '--write'):
        write = arg
    else:
        print "Uh oh, something went wrong. Try again, and read the docs! Use forage.py -h if you need help."
        sys.exit(2)

# Coming soon!
# if (write == "yaml" or "YAML"):
#     import yaml

print " Opening settings from settings.json ..."

# We open settings.json and load it to settings

with open('settings.json') as settings_file:
    settings = json.load(settings_file)

# This enables hydrometrics

if (settings["global"]["hydrometric"]["parse"] == "enabled"):
    output["hydrometric"] = {}
    data["hydrometric"]= {}
    hydrometric = True

print " Settings configured!"

print " Fetching data..."

# This iterates through all endpoints and fetches the data.

for endpoint in settings["endpoints"]:
    if (settings["endpoints"][endpoint]["type"] == "hydrometric" and hydrometric == True):
        endpoint = settings["endpoints"][endpoint]
        data["hydrometric"][endpoint["id"]] = []
        url = endpoint["root"] + endpoint["type"] + "/csv/" + endpoint["province"] + "/" + endpoint["timescale"] + "/" + endpoint["province"] + "_" + endpoint["id"] + "_" + endpoint["timescale"] + "_hydrometric.csv"
        with closing(requests.get(url, stream=True)) as r:
            lines = r.iter_lines()
            reader = csv.reader(lines, delimiter=',', quotechar='"')
            temp = False
            for row in reader:
                if (temp == False):
                    temp = True
                elif (row):
                    data["hydrometric"][endpoint["id"]].append([settings["endpoints"][endpoint["id"]]["name"], endpoint["id"], settings["endpoints"][endpoint["id"]]["timescale"], calendar.timegm(dateutil.parser.parse(row[1]).utctimetuple()), row[2], row[6]])

print " Data fetched!"

# If the write type is CSV, this function outputs to a CSV file called output-hydrometric.csv

if (write == "csv" or write == "CSV"):
    print " Writing data to CSV..."
    if (hydrometric == True):
        with open('output-hydrometric.csv', 'wb') as csvfile:
            writer = csv.writer(csvfile)
            writer.writerow(["Type", "Station Name", "Station ID", "Timescale", "Unix Timestamp", "Water Level (m)", "Discharge (m3/s)"])
            for dataset in data["hydrometric"]:
                for i in range(len(data["hydrometric"][dataset])):
                    row = data["hydrometric"][dataset][i]
                    if (row):
                        writer.writerow(["Hydrometric"] + row)
    print " Write to CSV complete!"

# If the write type is JSON, this function outputs to a JSON file called output-hydrometric.json

elif (write == "json" or write == "JSON"):
    print " Writing data to JSON..."
    if (hydrometric == True):
        with open('output-hydrometric.json', 'w') as outfile:
            json.dump(data, outfile)
    print " Write to JSON complete!"
# Coming soon: YAML support!
# elif (write == "yaml" or "YAML"):
#     if (hydrometric == True):
#         with open('output-hydrometric.yml', 'w') as outfile:
#             yaml.dump(data, outfile, default_flow_style=True)
print " Script complete!"
print " Shutting down..."

File: forage.rb
require 'optparse' # we use this to parse arguments/parameters
require 'time' # we need this to deal with dates and times
require 'csv' # we need this to read/write CSV files
require 'json' # we need this to read/write JSON files
require 'yaml' # we need this to read/write YAML files
require "net/http" # we need this to make http requests
require "uri" # we need this to make http requests

# Setup Variables

data = Hash.new()
write = false
hydrometric = false

# Parsing Command Line Arguments

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: forage.rb [options]"

  opts.on("-w", "--write TYPE", "Add output type, e.g. json or csv") do |w|
    write = w
  end

  opts.on("-h", "--help", "Get help!") do
    puts "Help coming soon!"
    exit
  end
end.parse!

# Parsing Settings

puts " Opening settings from settings.json ..."

settingsf = File.read('settings.json')
settings = JSON.parse(settingsf)

# Enabling hydrometrics

if (settings["global"]["hydrometric"]["parse"] == "enabled")
  data["hydrometric"] = Hash.new()
  hydrometric = true
end

puts " Settings configured!"

# Fetches Data

puts " Fetching data..."

settings["endpoints"].each do |key, value|
  temp = false
  data["hydrometric"][value["id"]] = []
  if value["type"] == "hydrometric" && hydrometric == true
    url = value["root"] + value["type"] + "/csv/" + value["province"] + "/" + value["timescale"] + "/" + value["province"] + "_" + value["id"] + "_" + value["timescale"] + "_hydrometric.csv"
    uri = URI.parse(url)
    http = Net::HTTP.new(uri.host, uri.port)
    request = Net::HTTP::Get.new(uri.request_uri)
    response = http.request(request)
    CSV.parse(response.body) do |row|
      if row
        if temp == false
          temp = true
        else
          data["hydrometric"][key].push([value["name"], value["id"], value["timescale"], Time.parse(row[1]).to_i, row[2], row[6]])
        end
      end
    end
  end
end

puts " Data fetched!"

# If CSV is enabled, writes to CSV

if write == "csv" || write == "CSV"
  puts " Writing data to CSV..."
  if hydrometric == true
    CSV.open("output-hydrometric.csv", "wb") do |csv|
      csv << ["Type", "Station Name", "Station ID", "Timescale", "Unix Timestamp", "Water Level (m)", "Discharge (m3/s)"]
      data["hydrometric"].each do |key, value|
        value.each do |k|
          csv << k.unshift("Hydrometric")
        end
      end
    end
  end
  puts " Write to CSV complete!"
end

# If JSON is enabled, writes to JSON

if write == "json" || write == "JSON"
  puts " Writing data to JSON..."
  if hydrometric == true
    File.open("output-hydrometric.json","w") do |f|
      f.write(data.to_json)
    end
  end
  puts " Write to JSON complete!"
end

# If YAML is enabled, writes to YAML

if write == "yaml" || write == "yml" || write == "YAML" || write == "YML"
  puts " Writing data to YAML..."
  if hydrometric == true
    File.open("output-hydrometric.yaml","w") do |f|
      f.write(data.to_yaml)
    end
  end
  puts " Write to YAML complete!"
end


puts " Script complete!"
puts " Shutting down..."
